Settings redesign

shobhitagarwal1612 opened the pull request:
In continuation of #1066

shobhitagarwal1612 committed:
Groups user access control in admin settings

shobhitagarwal1612 committed:
Adds summary and string resources

shobhitagarwal1612 committed:
Groups general settings and adds icons

shobhitagarwal1612 committed:
Refactors code
Removes redundant code

shobhitagarwal1612 committed:
Updates code for hiding user settings items

shobhitagarwal1612 committed:
Merge branch 'master' into settings-redesign

shobhitagarwal1612 committed:
Merges autosend wifi and cellular to a list preference

shobhitagarwal1612 committed:
Improved grouping in user interface

shobhitagarwal1612 committed:
Adds user and device identity category in general settings

shobhitagarwal1612 committed:
Renames preference titles

shobhitagarwal1612 committed:
Removes secondary summary from user access settings in admin settings

shobhitagarwal1612 committed:
Modifies the behaviour of platform preference

shobhitagarwal1612 committed:
Removes unused resources

shobhitagarwal1612 committed:
FIxes show/hide server preference from admin settings

shobhitagarwal1612 committed:
Deletes unused code

shobhitagarwal1612 committed:
Adds license

shobhitagarwal1612 committed:
Adds new icons
Refactors DisabledPreferencesRemover class

shobhitagarwal1612 committed:
Renames the drawables more accurately

shobhitagarwal1612 committed:
Refactors code

shobhitagarwal1612 committed:
Merge branch 'master' into settings-redesign

# Conflicts:
#	collect_app/src/main/res/values/strings.xml

shobhitagarwal1612 committed:
Renames string use_device_language to use_phone_language

shobhitagarwal1612 committed:
Merge branch 'master' into settings-redesign

# Conflicts:
#	collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderList.java
#	collect_app/src/main/java/org/odk/collect/android/preferences/AboutPreferencesActivity.java
#	collect_app/src/main/java/org/odk/collect/android/preferences/AggregatePreferencesActivity.java
#	collect_app/src/main/java/org/odk/collect/android/preferences/GooglePreferencesActivity.java
#	collect_app/src/main/java/org/odk/collect/android/preferences/OtherPreferencesActivity.java
#	collect_app/src/main/java/org/odk/collect/android/preferences/PreferencesActivity.java

lognaturel commented:
Looking great! I hope merging the conflicts isn't too time-consuming.

Two things I noticed are missing: the actual splash screen image and the form metadata.

shobhitagarwal1612 committed:
Adds check for disabling preferenceGroups having defined child preferences

shobhitagarwal1612 committed:
Adds toolbar to ServerSettings

shobhitagarwal1612 committed:
Adds toolbar for nested preference screens

shobhitagarwal1612 commented:
@lognaturel Done

lognaturel commented:
Looking good! 

Fonts are pretty small. Are those the Material design standards? Could we just bump everything up by a couple of sp? I changed the font size to large at the device level and that feels very comfortable to me. Maybe those sizes could be the defaults? I'm willing to be talked out of this if you think sticking to standards and helping users change the font size at the device level is better.

Passwords should have a show/hide option. We can also file this as a separate issue and do it as the eye overlay which I think is now standard. Or perhaps dialogs should be handled separately. I notice the color of the dialog buttons are also inconsistent.

Speaking of color inconsistency! I sometimes get a mix of blue and teal on Nexus 5, API 22. Everything is consistently blue on my Moto G4, Android 7.0.0.
<img src="https://user-images.githubusercontent.com/967540/26911159-7790e568-4bc0-11e7-9037-42575ed6e355.png" width=250>

I haven't done a proper review of the code yet but one quick thing I noticed is that there is now only one key for the auto send setting. That's probably the right thing to do but the reason @yanokwa mentioned using the same preference keys is for backwards compatibility. We really don't want users not getting their forms as expected because autosend was turned off without them noticing. Also, if users have the separate settings as part of configuration QR codes, those should still work. You might want to look at `FormMetadataMigrator` for ideas on how to handle that.

shobhitagarwal1612 committed:
Adds backward compatibility for autosend setting

shobhitagarwal1612 commented:
@lognaturel : 
> Fonts are pretty small. Are those the Material design standards? Could we just bump everything up by a couple of sp? I changed the font size to large at the device level and that feels very comfortable to me. Maybe those sizes could be the defaults? I'm willing to be talked out of this if you think sticking to standards and helping users change the font size at the device level is better.

The font size of the entire app or just the settings screen?


> I haven't done a proper review of the code yet but one quick thing I noticed is that there is now only one key for the auto send setting. That's probably the right thing to do but the reason @yanokwa mentioned using the same preference keys is for backwards compatibility. We really don't want users not getting their forms as expected because autosend was turned off without them noticing. Also, if users have the separate settings as part of configuration QR codes, those should still work. You might want to look at FormMetadataMigrator for ideas on how to handle that.

Added backward compatibilty for importing from QRCode

lognaturel commented:
For the font size, I'm talking specifically about the settings screen for now.

shobhitagarwal1612 committed:
Fixes layout of show_qrcode_fragment

shobhitagarwal1612 committed:
Adds backward compatibility for importing settings when app starts

shobhitagarwal1612 committed:
Fixes font size of preferences

shobhitagarwal1612 committed:
modfies textSize form 17sp to 18sp in settings

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java:
```diff
@@ -685,6 +688,41 @@ private boolean loadSharedPreferencesFromFile(File src) {
+     * This method is to provide backward compatibility with v1.7.0 and below
+     * Autosend was originally set into separate wifi and cellular autosend settings
+     */
+    private void checkForBackwardCompatility(Map<String, ?> entries, Editor prefEdit) {
```
A method name that starts with `check` implies it's going to return a boolean. I would name this something like `migrateAutosendPrefKeys`.

shobhitagarwal1612 replied:
Fixed

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/preferences/DisabledPreferencesRemover.java:
```diff
@@ -66,32 +42,86 @@
+     * Checks whether the preferenceGroup actually has any child preferences defined
+     */
+    private boolean hasChildPrefs(String preferenceKey) {
+        String[] preferenceScreensWithNoChildren = {
```
Isn't this the list WITH children, not with NO children?

shobhitagarwal1612 replied:
Fixed

lognaturel reviewed (changes requested):
I don't see any code for doing the migration from two autosend preferences to one in the case of a normal update. To verify:
1. `git checkout v1.7.0` and install to device
1. check both "autosend with wifi" and "autosend with network"
1. checkout this branch and install to device
Now I should see "Wifi or cellular" selected for "Autosend" in the form management settings but I don't.

Another small thing I noticed is that if I uncheck everything in Admin Settings > User Settings, the Server category disappears but the other three categories are still there. Form management is clickable but empty.

Other than that, it looks ready to merge! The first one is a must-fix. The second we can file as a separate issue.

shobhitagarwal1612 committed:
Renames methods

shobhitagarwal1612 committed:
Fixes migration of autosend preferenes from settings file

shobhitagarwal1612 committed:
Adds autosend preference migrator

shobhitagarwal1612 commented:
> I don't see any code for doing the migration from two autosend preferences to one in the case of a normal update. To verify:
> 
> git checkout v1.7.0 and install to device
> check both "autosend with wifi" and "autosend with network"
> checkout this branch and install to device
> Now I should see "Wifi or cellular" selected for "Autosend" in the form management settings but I don't.

Done

shobhitagarwal1612 committed:
Fixes disabled preferences remover

shobhitagarwal1612 commented:
@lognaturel : 

> Another small thing I noticed is that if I uncheck everything in Admin Settings > User Settings, the Server category disappears but the other three categories are still there. Form management is clickable but empty.

I have fixed this. Although there is still a minute bug left. The form management is visible and clickable if all the preferences are hidden but it works if any one or more other preferences is/are unhidden. We can file a separate issue for fixing this.

shobhitagarwal1612 committed:
Merge branch 'master' into settings-redesign

# Conflicts:
#	collect_app/src/main/res/values-de/strings.xml
#	collect_app/src/main/res/values-es/strings.xml
#	collect_app/src/main/res/values-fr/strings.xml
#	collect_app/src/main/res/values-it/strings.xml
#	collect_app/src/main/res/values-ja/strings.xml
#	collect_app/src/main/res/values-km/strings.xml
#	collect_app/src/main/res/values-sv-rSE/strings.xml
#	collect_app/src/main/res/values-th-rTH/strings.xml
#	collect_app/src/main/res/values-uk/strings.xml
#	collect_app/src/main/res/values/strings.xml

shobhitagarwal1612 committed:
Removes unused resources

lognaturel commented:
I verified that when I check check both auto send settings in v1.7.0, install this branch, then auto send is set to Wifi & Cellular as it should be.

However, when I checked auto send with network in v1.7.0, exported the settings file (Admin settings > menu > Save Settings to Disk), moved the generated file to /sdcard/odk/collect.settings, installed this branch and then launched it, I saw the "settings successfully imported" toast but auto send was still set to "Off" (It should have said "cellular"). This will likely be addressed by #1133 but I also wanted to keep a record of it here.
